File: Client/src/main/java/com/rspsi/editor/model/InstanceGeneratedHeightProvider.java
package com.rspsi.editor.model;

/** Supplies opcode-0 heights when a source terrain tile is replayed in an instance. */
@FunctionalInterface
public interface InstanceGeneratedHeightProvider {
    int heightAt(int sourceWorldX, int sourceWorldY);

    static InstanceGeneratedHeightProvider required() {
        return (x, y) -> {
            throw new IllegalStateException(
                    "Instance terrain contains generated heights; provide an instance height provider");
        };
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/model/InstanceGeneratedHeightProvider.kt
package com.rspsi.editor.model

/**
 * Supplies opcode-0 terrain heights when a source tile is replayed into an instance.
 *
 * This remains a single-abstract-method contract so existing Java and Kotlin lambda call sites
 * keep working without adapters. Height generation stays cache-independent at this boundary.
 */
@java.lang.FunctionalInterface
fun interface InstanceGeneratedHeightProvider {
    fun heightAt(
        sourceWorldX: Int,
        sourceWorldY: Int,
    ): Int

    companion object {
        /**
         * Returns the compatibility provider used when generated heights are not configured.
         *
         * It fails only when invoked, preserving the previous lazy error behavior.
         */
        @JvmStatic
        fun required(): InstanceGeneratedHeightProvider =
            InstanceGeneratedHeightProvider { _, _ ->
                throw IllegalStateException(
                    "Instance terrain contains generated heights; provide an instance height provider",
                )
            }
    }
}
